src/modules/flow/am2315/am2315.c

```diff
@@ -0,0 +1,224 @@
+    SOL_PTR_VECTOR_FOREACH_IDX (&devices, device, i) {
+        if (device->bus == bus && device->slave == slave) {
+            device->refcount++;
+            return device;
+        }
+    }
+
+    i2c = sol_i2c_open(bus, SOL_I2C_SPEED_10KBIT);
+    if (!i2c) {
+        SOL_WRN("Failed to open i2c bus");
+        return NULL;
+    }
+
+    device = calloc(1, sizeof(struct am2315));
+    if (!device)
+        goto fail;
+
+    device->i2c = i2c;
+    device->bus = bus;
+    device->slave = slave;
+
+    sol_ptr_vector_append(&devices, device);
+
+    return device;
+
+fail:
+    sol_i2c_close(i2c);
+
+    return NULL;
+}
+
+void
+am2315_close(struct am2315 *device)
+{
+    struct am2315 *itr;
+    int i;
+
+    device->refcount--;
+    if (device->refcount) return;
+
+    sol_i2c_close(device->i2c);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&devices, itr, i) {
+        if (itr == device) {
+            sol_ptr_vector_del(&devices, i);
+            break;
+        }
+    }
+
+    free(device);
+}
+
+static uint16_t
+_crc16(uint8_t *ptr, uint8_t len)
+{
+    uint16_t crc = 0xffff;
+    uint8_t i;
+
+    while (len--) {
+        crc ^= *ptr++;
+        for (i = 0; i < 8; i++) {
+            if (crc & 0x01) {
+                crc >>= 1;
+                crc ^= 0xa001;
+            } else {
+                crc >>= 1;
+            }
+        }
+    }
+
+    return crc;
+}
+
+static bool
+_update_readings(struct am2315 *device)
+{
+    time_t current_time = time(NULL);
+    uint8_t write_msg[] = { AM2315_READ_REG, AM2315_HUMIDITY_HIGH, AM2315_READ_LEN };
+    uint8_t read_data[8];
+    ssize_t read_data_size;
+    uint16_t crc;
+
+    if (current_time == -1) {
+        SOL_WRN("Could not get current time");
+        return false;
+    }
+
+    // TODO Should we care if time is changed backwards?
+    if (current_time - device->last_reading <= AM2315_INTERVAL_BETWEEN_READINGS) {
+        return true;
+    }
+
+    device->last_reading = current_time;
+
+    if (!sol_i2c_set_slave_address(device->i2c, device->slave)) {
+        SOL_WRN("Failed to set slave at address 0x%02x", device->slave);
+        return false;
+    }
+
+    // Write a message to read data
+    if (!sol_i2c_write(device->i2c, write_msg, sizeof(write_msg))) {
+        SOL_WRN("Could not read sensor");
+        return false;
+    }
+
+    // Datasheet asks for some delay
+    usleep(2000);
+
+    // Read 8 bytes: 1st is the function code, 2nd is data lenght,
+    // 3rd and 4th are humidity hi/lo, 5th and 6th are temperature
+    // hi/lo, 7th and 8th are CRC code to validade data.
+    read_data_size = sol_i2c_read(device->i2c, read_data, sizeof(read_data));
+
+    if (read_data_size != sizeof(read_data)) {
+        SOL_WRN("Could not read sensor data");
+        return false;
+    }
+
+    // CRC from first byte to 6th. iow, do not calculate CRC of CRC bytes
+    crc = _crc16(read_data, 6);
+    if (read_data[0] != AM2315_READ_REG || read_data[1] != AM2315_READ_LEN
```

Comment:
I'd rather split the checks in two blocks: the CRC verification (right after the calculation), and then checking if it's a read with size. This is slightly more readable (one-line conditions are nicer), and the error message can be more contextual ("CRC16 mismatch", etc).